feat(mechanics): Make stray projectiles not hit disabled fighters

tibetiroka opened the pull request:
**Feature**

This PR implements the feature suggested by @Amazinite during the last campfire chat.

## Summary
This PR makes it so fighters/drones don't get hit by stray projectiles when disabled. Specifically:
- When disabled, all projectiles targeting them clear their target
- Projectiles not specifically targeting a disabled fighter/drone will not hit it
  - Blast damage will still damage disabled fighters/drones
  - Proximity-triggered weapons (`trigger radius`) will still be triggered by the fighter/drone

Effectively, this means that most projectiles en route will not hit and destroy disabled fighters/drones, greatly enhancing their survivability.

## Testing Done
- [x] Fighters get disabled
- [x] Disabled fighters are not hit by random fire
- [x] Disabled fighters are hit by targeted fire

## Performance Impact
Should be negligible.

Hurleveur commented:
Please make this only happen for player fighters or we'll end up with many annoying disabled fighters everywhere

tibetiroka commented:
The devs specifically said that this would have to be common to all fighters. I'm about to do some real playtesting to get an idea of how annoying the fighters get; I'll have some results in a couple of hours.

Amazinite commented:
This is actually a lot easier than I was originally anticipating. 👀
I guess the collision set refactor from #6276 helped out a fair bit.

EjoThims commented:
'When disabled, all projectiles targeting the ship clear their target
Projectiles not specifically targeting the ship will not hit it'

Does this apply to all "ships" or only to "fighters/drones"?

If it applies to everything, this is a huge buff to looting/capping play styles.

tibetiroka commented:
The 'ship' is a drone/fighter in every case, I didn't change the mechanics for any other type. Sorry for the confusion.

tibetiroka commented:
Okay, here are my impressions:

For mining, this feels great. In systems with occasional pirate spawns my drones no longer die instantly. This is great, because now I can mine in some safer systems in the South without having to go back to the core for more mining drones. I have 3 Sunders and 6 mining drones, and now a single Fury doesn't mess up my day. Anything more serious still does enough damage to my drones to make mining unviable (because the drones have to be picked up, and a full Sunder isn't fast enough to catch most asteroids).

In combat, I tested a stock carrier build against various other ships. It seems that the number and type of missiles used is what really matters.

Against other enemies without missiles, or only weak missiles like javelins, most (or often all) drones/fighters survive the battle in a disabled state. However, against heavier weapons like torpedoes, they tend to still die in great numbers. Using the carrier against a stock Dreadnought with Torpedoes kills about half of them, and that's a certain victory for the carrier. Firelights tend to clear out nearly every drone/fighter, as they have a trigger radius.

Interestingly, the Solifuge with its barrage of trackers kills pretty much every drone and fighter. I think this is because multiple projectiles can damage the ship in one frame, even if it gets disabled. While this might be a bug, I think it actually helps the balance somewhat.

If someone has a save file for major FW battles, I could test how many fighters are left there. I couldn't find any snapshots amongst my own saves, unfortunately.

tibetiroka commented on source/Engine.cpp:
```diff
@@ -2738,3 +2738,16 @@ void Engine::EmplaceStatusOverlay(const shared_ptr<Ship> &it, Preferences::Overl
+	for(const auto &event : events)
+		if(event.Type() ^ ShipEvent::DISABLE)
+			for(auto &projectile : projectiles)
+				if(projectile.Target() == event.Target().get())
```
Oops. Should be fixed now.

Amazinite commented on source/Engine.cpp:
```diff
@@ -2738,3 +2738,16 @@ void Engine::EmplaceStatusOverlay(const shared_ptr<Ship> &it, Preferences::Overl
+	for(const auto &event : events)
+		if(event.Type() ^ ShipEvent::DISABLE)
+			for(auto &projectile : projectiles)
+				if(projectile.Target() == event.Target().get() && projectile.Target()->CanBeCarried())
```
Chrck if the event target is carriable outside of the projectile loop.

Amazinite commented on source/Engine.cpp:
```diff
@@ -2738,3 +2738,16 @@ void Engine::EmplaceStatusOverlay(const shared_ptr<Ship> &it, Preferences::Overl
+void Engine::HandleEvents()
+{
+	for(const auto &event : events)
+		if(event.Type() ^ ShipEvent::DISABLE && event.Target()->CanBeCarried())
```
I believe we want a bitwise and here, not a xor.
```suggestion
		if(event.Type() & ShipEvent::DISABLE && event.Target()->CanBeCarried())
```

tibetiroka replied:
Okay I'm genuinely getting worried about how much I'm getting wrong here. I should not be coding with a fever.

tibetiroka commented:
Well, at least that explains why missile barrages were still killing my fighters... They have a much higher survival rate now. Might even be too high for comfort.

tibetiroka commented:
Some realistic testing in FW battles using Warp's saves:

Continuous:
 - Defend Cebalrai: 3 surviving combat drones
 - Capture Cebalrai: 1 drone + 1 Lance
 - Attack Kaus Borealis: 6 drones + 1 Lance
 - Liberating Delta Sagittarii: 2 drones

With these changes:
 - Defend Cebalrai: 10 drones + 2 Lances
 - Capture Cebalrai: 4 Lances
 - Attack Kaus Borealis: 10 drones + 3 Lances
 - Liberating Delta Sagittarii: 5 drones + 1 Lance

I've only done these tests once and there seems to be a lot of variance.

Hurleveur commented:
I honestly feel like even with how things currently are there are often too many enemy fighters left over (tho the boarding preference changes alleviated the issue I have with that)

tibetiroka commented:
Honestly, even with these changes it doesn't feel like a 'fighter graveyard' to me... sure, there are more things to cycle through if you want to cap something, but it's not too bad.

Amazinite commented on source/Engine.cpp:
```diff
@@ -2738,3 +2738,16 @@ void Engine::EmplaceStatusOverlay(const shared_ptr<Ship> &it, Preferences::Overl
+
+void Engine::HandleEvents()
+{
+	for(const auto &event : events)
+		if(event.Type() & ShipEvent::DISABLE && event.Target()->CanBeCarried())
+			for(auto &projectile : projectiles)
+				if(projectile.Target() == event.Target().get())
+					projectile.BreakTarget();
```
We should probably change this so that it only iterates over each collection once instead of iterating over the projectiles list every time.
1. Iterate through all events to find the ships we need to break targeting on. (Throw `event.Target()` into a `set<>`.)
2. Iterate through all the projectiles and call BreakTarget on those that have a target in the above set. (`set.contains(projectile.Target())` or something like that.)

Realistically you'd only get multiple disable events in the same frame with bigger battles, but it'd be cleaner to set it up this way should we ever have to handle more events.

tibetiroka replied:
Even in bigger battles, a ship should only get disabled once in a frame, so we don't need it to be a `set<>`, but yeah, that would be cleaner.

Amazinite replied:
True. A vector would work just fine.

tibetiroka replied:
I just realized that a set is still better because we are doing lookups...

Amazinite replied:
And I realized it's `count`, not `contains`.  That's what I get for regularly swapping between three programming languages.

Amazinite commented on source/Engine.h:
```diff
@@ -171,6 +171,8 @@ class Engine {
 	void EmplaceStatusOverlay(const std::shared_ptr<Ship> &ship, Preferences::OverlayState overlaySetting, int value);
 
+	// Handle any events the engine is interested in.
+	void HandleEvents();
```
```suggestion
	void HandleEvents();

```

Hurleveur reviewed (commented):
Looks good code wise but it'd be nice to make this a gamerule (and maybe include there if it applies only to player ships or not, too)
I'd like the default to be it only applies to them but as players test this we'll be sure to get feedback and could adjust it then

tibetiroka commented:
Yeah this does need more testing by balance-oriented people

Amazinite reviewed (commented):
Heads up that I realized that since ships normally only change their targeting every half second, a ship may still fire at a fighter it was targeting and just disabled for another half, and those projectiles won't have their targeting broken with the ships since they were fired after the disable event.
https://github.com/endless-sky/endless-sky/blob/941639dc6f7a2abdb945a7b45740e77f9dbc9c32/source/AI.cpp#L673-L681
We could perhaps change this by making it so that targeting with disabled fighters is broken instantly by all ships instead of only having `disables` personality ships break targeting with disabled ships. So something like `(target->IsDisabled() && (personality.Disables() || target->CanBeCarried()))`

tibetiroka commented on source/Engine.cpp:
```diff
@@ -2738,3 +2738,18 @@ void Engine::EmplaceStatusOverlay(const shared_ptr<Ship> &it, Preferences::Overl
+
+void Engine::HandleEvents()
+{
+	set<const Ship*> disabledFighters;
```
No, things inside templates are ignored.

tibetiroka commented:
I added your change and also fixed collision checks for long beams (which is a really important fix), so fighters now have really good survivability.

I tested with 20 boxwings against a marauder leviathan (because that doesn't have missiles); previously, about three boxwings were destroyed, now they all survive.

Missiles are still dangerous, and any kind of non-disabling personality will still finish them off, but this is at a point where using fighters is actually sustainable.

Good thing is, we now have a whole lot of knobs to fine-tune the balance of this feature.

Amazinite commented on source/Engine.cpp:
```diff
@@ -2742,3 +2742,18 @@ void Engine::EmplaceStatusOverlay(const shared_ptr<Ship> &it, Preferences::Overl
+	for(const auto &event : events)
+		if(event.Type() & ShipEvent::DISABLE && event.Target()->CanBeCarried())
+			disabledFighters.insert(event.Target().get());
+	for(auto &projectile : projectiles)
+		if(disabledFighters.count(projectile.Target()))
+			projectile.BreakTarget();
```
No use in iterating over every projectile every frame if we're not doing anything with them.
```suggestion
	if(!disabledFighters.empty())
		for(auto &projectile : projectiles)
			if(disabledFighters.count(projectile.Target()))
				projectile.BreakTarget();
```

tibetiroka replied:
> Do we even have to use events? Projectiles already break their targetting in a couple of cases, so we could just add a check for disabled fighter to the conditions (in Projectiles::Move).

Wouldn't that have worse performance? That would be like iterating over every projectile every frame, which is worse than only checking the events.q

quyykk replied:
We already have to iterate over every projectile every frame to Move() them.

quyykk replied:
huh yeah I guess so whoops mb, misunderstood the intent.

quyykk commented on source/CollisionSet.cpp:
```diff
@@ -221,6 +221,12 @@ const vector<Collision> &CollisionSet::Line(const Point &from, const Point &to,
 			const Government *iGov = it->body->GetGovernment();
 			if(it->body != target && iGov && pGov && !iGov->IsEnemy(pGov))
 				continue;
+			if(collisionType == CollisionType::SHIP)
+			{
+				Ship *ship = reinterpret_cast<Ship *>(it->body);
+				if(ship != target && ship->CanBeCarried() && ship->IsDisabled())
+					continue;
+			}
```
This seems like the wrong place to put this in. I think the logic should be higher up, i.e. actively ignoring a collision hit that was found instead of changing the collision detection code to not find a collision.

Amazinite replied:
In that case, the `all` boolean parameter of this function would need to always be true.

Amazinite replied:
I'll open a PR to this branch to make the necessary CollisionSet changes, because I do actually agree that it's not the best to make this CollisionSet's responsibility.

Amazinite replied:
https://github.com/tibetiroka/endless-sky/pull/2

quyykk commented on source/CollisionSet.cpp:
```diff
@@ -295,6 +301,12 @@ const vector<Collision> &CollisionSet::Line(const Point &from, const Point &to,
 			const Government *iGov = it->body->GetGovernment();
 			if(it->body != target && iGov && pGov && !iGov->IsEnemy(pGov))
 				continue;
+			if(collisionType == CollisionType::SHIP)
+			{
+				Ship *ship = reinterpret_cast<Ship *>(it->body);
+				if(ship != target && ship->CanBeCarried() && ship->IsDisabled())
+					continue;
+			}
```
This and a couple of lines above are duplicated, make it a lambda maybe? Or a local function.

tibetiroka commented:
I haven't gone through all the saves yet, but after the fixes I now get:

 - Defend Cebalrai: 5 lances + 5 drones
 - Liberating Delta Sagittarii: 2 lances + 7 drones

quyykk commented on source/CollisionSet.cpp:
```diff
@@ -201,6 +201,21 @@ const vector<Collision> &CollisionSet::Line(const Point &from, const Point &to,
 	Closest closer_result(1.);
 	lineResult.clear();
 
+	const auto canHit = [&](vector<Entry>::const_iterator &it){
```
```suggestion
	const auto CanHit = [this, pGov, target](vector<Entry>::const_iterator &it) {
```

quyykk commented on source/CollisionSet.cpp:
```diff
@@ -216,10 +231,7 @@ const vector<Collision> &CollisionSet::Line(const Point &from, const Point &to,
-			// projectile or the object has no government, it will always hit.
-			const Government *iGov = it->body->GetGovernment();
-			if(it->body != target && iGov && pGov && !iGov->IsEnemy(pGov))
+			if(!canHit(it))
```
```suggestion
			if(!CanHit(it))
```

quyykk commented on source/CollisionSet.cpp:
```diff
@@ -290,10 +302,7 @@ const vector<Collision> &CollisionSet::Line(const Point &from, const Point &to,
-			// projectile or the object has no government, it will always hit.
-			const Government *iGov = it->body->GetGovernment();
-			if(it->body != target && iGov && pGov && !iGov->IsEnemy(pGov))
+			if(!canHit(it))
```
```suggestion
			if(!CanHit(it))
```

quyykk commented on source/Engine.cpp:
```diff
@@ -2742,3 +2742,19 @@ void Engine::EmplaceStatusOverlay(const shared_ptr<Ship> &it, Preferences::Overl
+
+void Engine::HandleEvents()
+{
+	set<const Ship *> disabledFighters;
+	for(const auto &event : events)
+		if(event.Type() & ShipEvent::DISABLE && event.Target()->CanBeCarried())
+			disabledFighters.insert(event.Target().get());
+	if(!disabledFighters.empty())
+		for(auto &projectile : projectiles)
+			if(disabledFighters.count(projectile.Target()))
+				projectile.BreakTarget();
```
comment plz

tibetiroka replied:
```suggestion
	// When fighters are disabled, all projectiles targeting them should seek out a new target.
	set<const Ship *> disabledFighters;
	for(const auto &event : events)
		if(event.Type() & ShipEvent::DISABLE && event.Target()->CanBeCarried())
			disabledFighters.insert(event.Target().get());
	if(!disabledFighters.empty())
		for(auto &projectile : projectiles)
			if(disabledFighters.count(projectile.Target()))
				projectile.BreakTarget();
```

quyykk reviewed (commented):
Actually, looking at it a bit more closely you could do it in Projectile::Move like how it is done with `targetGovernment`: Store the state of the target when the projectile is launched/created, and if the original state is enabled and the new state is disabled, then break the target.

tibetiroka commented:
> Actually, looking at it a bit more closely you could do it in Projectile::Move like how it is done with `targetGovernment`: Store the state of the target when the projectile is launched/created, and if the original state is enabled and the new state is disabled, then break the target.

Honestly, that sounds like overcomplicating it compared to just using the events

RisingLeaf commented:
+1 to what quyykk suggests

quyykk commented on source/Projectile.cpp:
```diff
@@ -82,6 +85,8 @@ Projectile::Projectile(const Projectile &parent, const Point &offset, const Angl
 	hitsRemaining = weapon->PenetrationCount();
 
 	cachedTarget = TargetPtr().get();
+	if(cachedTarget)
+		targetDisabled = cachedTarget->IsDisabled();
```
This should be using the `targetDisabled` value of the parent.

tibetiroka replied:
What do you mean? I thought you wanted me to cache whether the target was disabled, and if it changes to disabled, break the target. I don't think how the parent comes into that.

quyykk replied:
It's cause this is the constructor for a submunition projectile. I meant the parent projectile, so like `targetDisabled = parent.targetDisabled;` (like it's done for the targetGovernment too).

tibetiroka replied:
aaaah that makes sense

Amazinite reviewed (commented):
We still only want to break targeting with fighters.

quyykk commented on source/AI.cpp:
```diff
@@ -675,7 +675,8 @@ void AI::Step(Command &activeCommands)
 			targetTurn = (targetTurn + 1) & 31;
-			if(targetTurn == step || !target || target->IsDestroyed() || (target->IsDisabled() && personality.Disables())
+			if(targetTurn == step || !target || target->IsDestroyed() || (target->IsDisabled() &&
+					(personality.Disables() || target->CanBeCarried()))
```
```suggestion
					(personality.Disables() || (target->CanBeCarried() && !personality.IsVindictive())))
```

idk about this one, maybe?

tibetiroka replied:
I'll leave this one up to the balance folks

Amazinite replied:
Vindictive is a rarely used personality, so it's not really a balance concern, more a thematic concern with how vindictive is supposed to behave.
> vindictive: this ship continues to fire at its target even after the target is dead, while the target is beginning to explode.

eljaredoe commented:
I would like to ask to please get this as a plugin. I've been hoping for this change since the last two campfire discussions lol. I don't know how to implement this myself and it would absolutely make my week if it isn't too much trouble.

If it is too much trouble, where can I find out how to add this to my game?

tibetiroka commented:
You cannot use this as a plugin, unfortunately. You can still get access to this feature via ESLauncher2 or ESManager by using this pull request for an instance. (You can find more info on ESLauncher2 in the discord server, and ESManager should be pretty straightforward from there if you are on linux.)

Tecmes commented:
As pointed out early, this creates multiple disabled drones when facing navy. Then it becomes even harder to board anything else. It is said above that the devs wanted this feature to affects both player and AI, but I don't understand why?

Hurleveur commented:
I agree, honestly there were too many drones already for me, and this can only make that worse.
That said, the mixed boarding value feature does make it more bearable, but it remains annoying

Definitely worth opening an issue to at least get the option to turn it off for the AI, I think.

quyykk commented:
Would prioritizing non-drones when pressing Board be a possible solution? No idea if this is a good idea though.

Saugia commented:
>As pointed out early, this creates multiple disabled drones when facing navy. Then it becomes even harder to board anything else. It is said above that the devs wanted this feature to affects both player and AI, but I don't understand why?

I had mentioned that we feel it odd to give the player advantages or mechanics that we wouldn't give the AI. This isn't 100% the case, but for something like disabled status, it would be unfair if the player's ships are safer than the AI's.

>Then it becomes even harder to board anything else.

I don't think this is good enough reason against the above to not implement it. Boarding isn't the primary content of the game, and isn't much of a concern versus ship status and description. Typically you'll be visually selecting and boarding vessels. I get that hitting B is easier, but, well, it makes sense that the more ships there are, the more you have to navigate them.

>Would prioritizing non-drones when pressing Board be a possible solution? No idea if this is a good idea though.

Now, perhaps we could consider priorities in boarding. I'm not sure what the others think about such a code though, and personally, like the above, I don't feel like this warrants something like a code enhancement for easier boarding. Having a cycling boarding key, or being able to cycle targets/nearest targets, seem more than appropriate for navigating this situation.

Hurleveur commented:
> Would prioritizing non-drones when pressing Board be a possible solution? No idea if this is a good idea though.

If you use the value preference, or even mixed, thats pretty much what it does and why I coded it (tho it goes beyond drones ofc)

43897689431805741092 commented:
PLEASE make this disable-able with a gamerule or something.
(I would just go off and quietly build the game myself while excluding this particular PR but can't figure that out.)
